### PR TITLE
Implement MaxDispatches in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
The MaxDispatches implementation is largely from #2598.

This PR also adds a new Ice/maxDispatches test, and refactors the implementation of the idle check.

### CODE REVIEW

File: csharp/test/Ice/maxDispatches/AllTests.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        Test.TestIntfPrx pMax1 = Test.TestIntfPrxHelper.createProxy(communicator, proxyStringMax1);
+
+        string proxyStringSerialize = $"test: {helper.getTestEndpoint(3)}";
+        Test.TestIntfPrx pSerialize = Test.TestIntfPrxHelper.createProxy(communicator, proxyStringSerialize);
+
+        await testMaxDispatches(p, 100, helper.getWriter());
+        await testMaxDispatches(pMax10, 10, helper.getWriter());
+        await testMaxDispatches(pMax1, 1, helper.getWriter());
+
+        // Serialize does not limit dispatches with "true" AMD.
```
Comment (bernardnormier) [inline csharp/test/Ice/maxDispatches/AllTests.cs:28]: This test shows that with AMD, xxx.ThreadPool.Serialize=1 does not limit concurrent dispatches.

I am in favor of removing this Serialize property / implementation. I don't find it useful now that we have MaxDispatches.

File: csharp/src/Ice/Internal/UdpTransceiver.cs [hunk 2]
```diff
@@ -9,15 +9,6 @@ namespace Ice.Internal;
 
 internal sealed class UdpTransceiver : Transceiver
 {
-    public bool isWaitingToBeRead
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/UdpTransceiver.cs:12]: Removed all isWaitingToBeRead.

File: csharp/src/Ice/ConnectionI.cs [hunk 3]
```diff
@@ -1440,42 +1442,28 @@ internal ConnectionI(
...
+    internal void idleCheck(TimeSpan idleTimeout)
     {
         lock (this)
         {
-            if (_state == StateActive || _state == StateHolding)
+            if (_state == StateActive && _idleTimeoutTransceiver!.idleCheckEnabled)
             {
                 int idleTimeoutInSeconds = (int)idleTimeout.TotalSeconds;
 
-                if (_transceiver.isWaitingToBeRead)
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:1453]: Remove first if ... block

File: csharp/src/Ice/ConnectionI.cs [hunk 4]
```diff
@@ -1440,42 +1442,28 @@ internal ConnectionI(
...
-    /// <summary>Aborts the connection with a <see cref="ConnectionAbortedException" /> if the connection is active or
-    /// holding.</summary>
-    internal void idleCheck(TimeSpan idleTimeout, Action rescheduleTimer)
+    /// <summary>Aborts the connection with a <see cref="ConnectionAbortedException" /> if the connection is active and
+    /// does not receive any byte for some time. See the IdleCheckTransceiverDecorator.</summary>
+    internal void idleCheck(TimeSpan idleTimeout)
     {
         lock (this)
         {
-            if (_state == StateActive || _state == StateHolding)
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:1449]: When the state is StateHolding, we're not reading off the network, so we should not "idle check". The difficulty is when we resume to StateActivate, we want to re-enable the idle check / read timer.

File: csharp/src/Ice/ConnectionI.cs [hunk 5]
```diff
@@ -1655,31 +1643,42 @@ private void setState(int state)
...
                     }
+
+                    if (_maxDispatches > 0 && _dispatchCount >= _maxDispatches)
+                    {
+                        // Don't resume reads if we're at or over the _maxDispatches limit.
+                        return;
+                    }
+
                     _threadPool.register(this, SocketOperation.Read);
+                    _idleTimeoutTransceiver?.enableIdleCheck();
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:1660]: We now enable/disable the idle check when we switch to StateActive resp. StateHolding.

File: csharp/src/Ice/Internal/IdleTimeoutTransceiverDecorator.cs [hunk 6]
```diff
@@ -60,7 +62,6 @@ public int initialize(Buffer readBuffer, Buffer writeBuffer, ref bool hasMoreDat
 
         if (op == SocketOperation.None) // connected
         {
-            rescheduleReadTimer();
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/IdleTimeoutTransceiverDecorator.cs:63]: Redundant with the new "enableIdleCheck()" called when switching to StateActive.

File: csharp/src/Ice/ConnectionI.cs [hunk 7]
```diff
@@ -1655,31 +1643,42 @@ private void setState(int state)
...
                     {
                         return;
                     }
-                    if (_state == StateActive)
+
+                    if (_maxDispatches > 0 && _dispatchCount >= _maxDispatches)
                     {
-                        _threadPool.unregister(this, SocketOperation.Read);
+                        // Reads are already disabled if the _maxDispatches limit is reached or exceeded.
+                        return;
```
Comment (externl) [inline csharp/src/Ice/ConnectionI.cs:1677]: Should this be a `break;`? Don't we still want to set `_state` to `StateHolding`?

File: csharp/test/Ice/maxDispatches/.gitignore
Comment (externl) [inline csharp/test/Ice/maxDispatches/.gitignore:1]: What is this file needed for?

File: csharp/src/Ice/ConnectionI.cs (hunk 7 above)
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs:1677]: Yes, we want to switch state. Fixed.

File: csharp/test/Ice/maxDispatches/.gitignore
Comment (bernardnormier) [reply csharp/test/Ice/maxDispatches/.gitignore:1]: I just copied it from another test.

**Review by externl — APPROVED** [review]
Looks good!

File: csharp/test/Ice/maxDispatches/.gitignore
Comment (externl) [reply csharp/test/Ice/maxDispatches/.gitignore:1]: It doesn't look like it should be needed.

File: csharp/test/Ice/maxDispatches/TestI.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    public override async Task opAsync(Current current)
+    {
+        lock (_mutex)
+        {
+            if (++_dispatchCount > _maxDispatchCount)
+            {
+                _maxDispatchCount = _dispatchCount;
+            }
+        }
+        await Task.Delay(TimeSpan.FromMilliseconds(50));
```
Comment (pepone) [inline csharp/test/Ice/maxDispatches/TestI.cs:22]: It would be best to not use a delay, for example use a monitor and signal it from an operation. This requires a separate OA.

**Review by pepone — APPROVED** [review]

File: csharp/test/Ice/properties/Client.cs [hunk 9]
```diff
@@ -116,6 +116,7 @@ public override void run(string[] args)
             catch (System.ArgumentException)
             {
             }
+            Console.Out.WriteLine("ok");
```
Comment (bernardnormier) [inline csharp/test/Ice/properties/Client.cs:119]: Small unrelated fix.

**Review by bentoi — APPROVED** [review]
Looks good. I still don't like much the circular dependency between the connection class and the idle timeout transceiver. In the end, it's more complicated with this decorator than just deal with the timers in the connection class.
